tmt/steps/provision/__init__.py

```diff
@@ -200,6 +462,30 @@ def start(self) -> None:
         """
         self.debug(f"Doing nothing to start guest '{self.guest}'.")
 
+    # A bit of Python magic: we want to run `sync()` when first accessing
```

Comment:
It is an interesting concept. I was thinking while reviewing, though, if it would not be more understandable if facts syncing would be explicit, instead of implicit as here?

It would improve the understandability and readability, and also it would be clear when the facts are synced?

Or is there a reason we want this nice way of doing it on first access. It is definitely cool (we do not need to care when we gather the facts), on the other hand it makes the readability of the code worse.

Reply:
It's possible to sync facts explicitly, the method is available if ever needed. For the initial sync, I opted for the on-demand approach, mostly because I didn't want to ask any part of the code to take responsibility of fetching the data for the first time. It cannot be done in `Guest.__init__()` as the hostname is still unknown, and when the guest is provisioned, either the `provision` step or each `provision` plugin would have to call this sync before letting the guest out of their sight. And it seemed to me that such a call is easy to forget, easy to drop in refactoring. Without the call, `Guest.facts` would be a black hole and source of tracebacks - I didn't want them to be `Optional`. Fetching data on the first access was fairly easy to do, there's no way `provision` machinery would forget to run the sync, and whoever is interested in refreshing the data, it's still possible. For `Guest.facts` users, the field would be always valid even if we'd cut away half of the provisioning step.

Reply:
Yeah, seems the benefits outweigh the more complex code to understand. Thanks